### PR TITLE
Synthesize InvalidPointerCheck call to take function parameters.

### PR DESCRIPTION
**Pull request by CristinaCristescu**
Creating the call for the cling invalid check function required that
the function have a const void *, thus a cast was injected.
Also renamed some of the variable to be more specific and
self-explanatory.

### CODE REVIEW

**Comment by CristinaCristescu** [comment]
I don't know if the names are yet self-explanatory? :-)
Cling test passes.
Ctest should be done soon!

**Comment by CristinaCristescu** [comment]
Ctest passes!

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 1]
```diff
@@ -91,32 +91,45 @@ namespace cling {
...
-              = m_Context.getTrivialTypeSourceInfo(Arg->getType(), Loc);
-      Expr* castExpr = m_Sema.BuildCStyleCastExpr(Loc, TSI, Loc, call).get();
 
-      return castExpr;
+      Expr* clingThrowIfInvalidPointerCallExpr
+        = m_Sema.BuildDeclarationNameExpr(CSS,
+                                          *m_clingthrowIfInvalidPointerCache,
+                                         /*ADL*/ false).get();
+      const clang::FunctionProtoType* clingThrowIfInvalidPointerType
+        = (const clang::FunctionProtoType*)(clingThrowIfInvalidPointerCallExpr->
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Can you do an `llvm::dyn_cast`, please?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 2]
```diff
@@ -91,32 +91,45 @@ namespace cling {
...
+                                                      getType().getTypePtr());
+
+      TypeSourceInfo* constVoidPtrTSI = m_Context.getTrivialTypeSourceInfo(
+        clingThrowIfInvalidPointerType->getParamType(2), Loc);
+
+      Expr* voidArg
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: That's a `voidPtrArg`, please :-)

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 3]
```diff
@@ -91,32 +91,45 @@ namespace cling {
...
-        = m_Sema.BuildDeclarationNameExpr(CSS, *m_LookupResult,
-                                         /*ADL*/ false).get();
-      Expr* call = m_Sema.ActOnCallExpr(S, unresolvedLookup, Loc,
-                                        args, Loc).get();
-      TypeSourceInfo* TSI
-              = m_Context.getTrivialTypeSourceInfo(Arg->getType(), Loc);
-      Expr* castExpr = m_Sema.BuildCStyleCastExpr(Loc, TSI, Loc, call).get();
 
-      return castExpr;
+      Expr* clingThrowIfInvalidPointerCallExpr
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Here it's okay to name it "checkCall" or similar. No need to be *that* verbose ;-)

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 4]
```diff
@@ -91,32 +91,45 @@ namespace cling {
...
+
+      Expr* voidArg
+        = m_Sema.BuildCStyleCastExpr(Loc, constVoidPtrTSI, Loc,
+                                     Arg).get();
+
+      Expr *args[] = {VoidSemaArg, VoidExprArg, voidArg};
+
+      Expr* call = m_Sema.ActOnCallExpr(S, clingThrowIfInvalidPointerCallExpr,
+                                         Loc, args, Loc).get();
+      if (call) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: This is usually spelled as
```
if (Expr* call = m_Sema.ActOnCallExpr(S, callExpr, Loc, args, Loc).get()) {
```

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 5]
```diff
@@ -91,32 +91,45 @@ namespace cling {
...
+        clang::TypeSourceInfo* argTSI = m_Context.getTrivialTypeSourceInfo(
+                                        Arg->getType(), Loc);
+        Expr* castExpr = m_Sema.BuildCStyleCastExpr(Loc, argTSI, Loc, call).get();
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: maybe just return the result of `m_Sema.BuildCStyleCastExpr...`

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 6]
```diff
@@ -146,17 +159,17 @@ namespace cling {
...
-      m_LookupResult = new LookupResult(m_Sema, Name, noLoc,
+      m_clingthrowIfInvalidPointerCache = new LookupResult(m_Sema, Name, noLoc,
                                         Sema::LookupOrdinaryName,
                                         Sema::ForRedeclaration);
-      m_Sema.LookupQualifiedName(*m_LookupResult,
+      m_Sema.LookupQualifiedName(*m_clingthrowIfInvalidPointerCache,
                                  m_Context.getTranslationUnitDecl());
-      assert(!m_LookupResult->empty() &&
+      assert(!m_clingthrowIfInvalidPointerCache->empty() &&
               "cling_runtime_internal_throwIfInvalidPointer");
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Can you state "lookup of cling_runtime_internal_throwIfInvalidPointer failed"?

Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Yes that make more sense, done!

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 5 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Ok

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 4 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Of course, sure!

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 3 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Yes, correct!

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 2 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: VoidPtrArg it is! It's true that voidArg can be misleading!

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp (hunk 1 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: Done!

**Comment by karies** [comment]
Thanks, merged as ef7dbb3!

**Pull request closed without merging**